py_trees/trees.py

```diff
@@ -192,21 +197,35 @@ def replace_subtree(self, unique_id, subtree):
                     return True
         return False
 
-    def setup(self, timeout):
+    def setup(self, timeout: float, visitor: visitors.VisitorBase=None):
         """
-         Relays to calling the :meth:`~py_trees.behaviour.Behaviour.setup` method
-         on the root behaviour. This in turn should get recursively called down through
-         the entire tree (if it does not, a behaviour with children has not
-         recursively called :meth:`~py_trees.behaviour.Behaviour.setup` on each of it's
-         children as it should).
+        Crawls across the tree calling :meth:`~py_trees.behaviour.Behaviour.setup`
+        on each behaviour.
+
+        Visitors can optionally be provided to provide a node-by-node analysis
+        on the result of each node's :meth:`~py_trees.behaviour.Behaviour.setup`
+        before the next node's :meth:`~py_trees.behaviour.Behaviour.setup` is called.
+        This is useful on trees with relatively long setup times to progressively
+        report out on the current status of the operation.
 
         Args:
-            timeout (:obj:`float`): time to wait (use common.Duration.INFINITE to block indefinitely)
+            timeout (:obj:`float`): time (s) to wait (use common.Duration.INFINITE to block indefinitely)
+            visitor (:class:`~py_trees.visitors.VisitorBase`): runnable entities on each node after it's setup
 
         Raises:
             Exception: be ready to catch if any of the behaviours raise an exception
         """
-        self.root.setup(timeout)
+        def visited_setup():
+            for node in self.root.iterate():
+                node.setup()
+                if visitor is not None:
+                    node.visit(visitor)
+        setup_process = multiprocessing.Process(target=visited_setup)
+        setup_process.start()
+        setup_process.join(timeout=timeout)
+        if setup_process.is_alive():
+            setup_process.terminate()
+            raise RuntimeError("tree setup() timed out")
```

Comment:
Consider TimeoutError instead. https://docs.python.org/3/library/exceptions.html#TimeoutError

Reply:
I've wanted to use that one in the past too. That however, is a subclass of OSError and the description explicitly states that it used for timeouts on system functions. It wouldn't kill anybody, but would be technically wrong to use it. If we did care, we should have our own TimeoutError as a custom exception, but since we've decided not to worry about these, I think lumping it in with RuntimeError is not going to cause anybody grief.